src/kudu/server/diagnostics_log.cc

```diff
@@ -0,0 +1,361 @@
+    log_dir_(std::move(log_dir)),
+    metric_registry_(metric_registry),
+    wake_(&lock_),
+    metrics_log_interval_(MonoDelta::FromSeconds(60)),
+    symbols_(new SymbolSet()) {
+}
+DiagnosticsLog::~DiagnosticsLog() {
+  Stop();
+}
+
+void DiagnosticsLog::SetMetricsLogInterval(MonoDelta interval) {
+  MutexLock l(lock_);
+  metrics_log_interval_ = interval;
+}
+
+void DiagnosticsLog::DumpStacksNow(std::string reason) {
+  MutexLock l(lock_);
+  dump_stacks_now_reason_ = std::move(reason);
+  wake_.Signal();
+}
+
+
+Status DiagnosticsLog::Start() {
+  unique_ptr<RollingLog> l(new RollingLog(Env::Default(), log_dir_, "diagnostics"));
+  RETURN_NOT_OK_PREPEND(l->Open(), "unable to open diagnostics log");
+  log_ = std::move(l);
+  Status s = Thread::Create("server", "diag-logger",
+                            &DiagnosticsLog::RunThread,
+                            this, &thread_);
+  if (!s.ok()) {
+    // Don't leave the log open if we failed to start our thread.
+    log_.reset();
+  }
+  return s;
+}
+
+void DiagnosticsLog::Stop() {
+  if (!thread_) return;
+
+  {
+    MutexLock l(lock_);
+    stop_ = true;
+    wake_.Signal();
+  }
+  thread_->Join();
+  thread_.reset();
+  stop_ = false;
+  WARN_NOT_OK(log_->Close(), "Unable to close diagnostics log");
+}
+
+MonoTime DiagnosticsLog::ComputeNextWakeup(DiagnosticsLog::WakeupType type) const {
+  switch (type) {
+    case WakeupType::STACKS:
+      if (FLAGS_diagnostics_log_stack_traces_interval_ms > 0) {
+        // Instead of directly using the configured interval, we use a uniform random
+        // interval whose mean is the configured value. This prevents biasing our stack
+        // samples. For example, if there is some background process which happens once a
+        // minute, and the user also configured the stacks to once a minute, an operator
+        // might incorrectly surmise that the background task was _always_ running.
+        // Randomizing the samples avoids such correlations.
+        Random rng(GetRandomSeed32());
+        int64_t ms = rng.Uniform(FLAGS_diagnostics_log_stack_traces_interval_ms * 2);
+        return MonoTime::Now() + MonoDelta::FromMilliseconds(ms);
+      } else {
+        // Stack tracing is disabled. However we still wake up periodically because the
+        // flag is runtime-modifiable, and we need to wake up to notice that it might have
+        // changed.
+        return MonoTime::Now() + MonoDelta::FromSeconds(5);
+      }
+      break;
+    case WakeupType::METRICS:
+      return MonoTime::Now() + metrics_log_interval_;
+  }
+  __builtin_unreachable();
+}
+
+void DiagnosticsLog::RunThread() {
+  MutexLock l(lock_);
+
+  // Set up a priority queue which tracks our future scheduled wake-ups.
+  typedef pair<MonoTime, WakeupType> QueueElem;
+  priority_queue<QueueElem, vector<QueueElem>, std::greater<QueueElem>> wakeups;
+  wakeups.emplace(ComputeNextWakeup(WakeupType::METRICS), WakeupType::METRICS);
```

Comment:
you are probably interested in metrics more than stacks.

Reply:
Let me remove the stacks in this commit.